backend/alert/models.py

```diff
@@ -322,6 +323,101 @@ class Registration(models.Model):
+        """
+        Returns a tuple of filters MIRRORING (not defining) the behavior of the
+        is_waiting_for_close property. Used for database filtering of registrations (you cannot
+        filter by a property value); unpack the filters with a single star
+        (NOT a double star like is_active_filter).
+        Example:
+            Registration.objects.filter(*Registration.is_waiting_for_close_filter())
+
+        As is the case with any Q filters, you must unpack these filters positionally before any
+        kwarg filters. If you are unfamiliar with Q queries, see the following docs:
+        https://docs.djangoproject.com/en/3.1/topics/db/queries/#complex-lookups-with-q-objects
+
+        All of these conditions are straightforward except for the last. The last condition
+        effectively guarantees the next registration in this registration's resubscribe chain
+        (if it exists) hasn't been deleted or cancelled. If the user has auto-resubscribe on and
+        then cancels the next registration, we don't want the previous registration to send
+        a close notification. Based on the way registrations are shown to the user in PCA,
+        they would expect cancelling or deleting the head of a resubscribe chain to halt all
+        notifications, close or otherwise, from that chain; this condition guarantees that.
+        Note that the only way for a registration 2 away from this one would be canceled is if
+        an alert was sent for the next registration, meaning a close notification for this
+        registration must have already been sent so close_notification_sent=True, or otherwise
+        the user cancelled and resubscribed multiple times, which would mean the next registration
+        would be cancelled. In all cases, is_waiting_for_close would correctly be False based
+        on these filters.
+        """
+        return (
+            Q(notification_sent=True),
+            Q(close_notification=True),
+            Q(deleted=False),
+            Q(cancelled=False),
+            Q(close_notification_sent=False),
+            (
+                Q(resubscribed_to__isnull=True)
+                | (
+                    Q(resubscribed_to__isnull=False)  # SQL short-circuit logic is not guaranteed
+                    & Q(resubscribed_to__cancelled=False)
+                    & Q(resubscribed_to__deleted=False)
+                )
+            ),
+        )
+
+    @property
+    def is_waiting_for_close(self):
+        """
+        True if the registration is waiting to send a close notification to the user
+        once the section closes.  False otherwise.
+        """
+
+        # WARNING: you should be frugal with your usage of this property, since it hits the DB
+        # each time it is called.
+
+        return Registration.objects.filter(
+            *Registration.is_waiting_for_close_filter(), id=self.id
+        ).exists()
+
+    @property
+    def last_notification_sent_at(self):
```

Comment:
The issue is, this is actually meant to avoid looking at the last in chain, because the last in chain usually hasn't sent a notification yet -- so for instance if someone has auto resubscribe on, `notification_sent_at` of the last in chain would always be null -- and this is what happens on the frontend (a registration shows no alerts sent yet, even if alerts have indeed been sent for that section). Also we have to think beyond just resubscribe chains because someone could create a new registration instead of resubscribing and they would still expect "last notification sent at" to be the last time a notification was sent about that section. But if the efficiency of constricting the search to resubscribe chains would be worth it to avoid that edge case, then maybe we can look at the most recent registration in the chain which has a non-null `notification_sent_at` value

Reply:
Ah I totally see, that makes a lot of sense. In that case though could you add a docstring explaining exactly what this is used for on the frontend? Would clear things up

Reply:
Yeah will do